presto-spark-base/src/main/java/com/facebook/presto/spark/planner/PrestoSparkRddFactory.java

```diff
@@ -331,23 +334,35 @@ else if (rddInputs.size() == 0) {
     }
 
     private PrestoSparkTaskSourceRdd createTaskSourcesRdd(
+            PlanFragmentId fragmentId,
             JavaSparkContext sparkContext,
             Session session,
             PartitioningHandle partitioning,
             List<TableScanNode> tableScans,
             Map<PlanNodeId, SplitSource> splitSources,
             Optional<Integer> numberOfShufflePartitions)
     {
-        ListMultimap<Integer, TaskSource> taskSourcesMap = ArrayListMultimap.create();
+        ListMultimap<Integer, SerializedPrestoSparkTaskSource> taskSourcesMap = ArrayListMultimap.create();
         for (TableScanNode tableScan : tableScans) {
             SplitSource splitSource = requireNonNull(splitSources.get(tableScan.getId()), "split source is missing for table scan node with id: " + tableScan.getId());
             List<ScheduledSplit> scheduledSplits = getSplitsAndCloseSource(tableScan, splitSource);
+            log.info("Found %s splits for table scan node with id %s", scheduledSplits.size(), tableScan.getId());
             shuffle(scheduledSplits);
             SetMultimap<Integer, ScheduledSplit> assignedSplits = assignSplitsToTasks(session, partitioning, scheduledSplits);
-            asMap(assignedSplits).forEach((partitionId, splits) ->
-                    taskSourcesMap.put(partitionId, new TaskSource(tableScan.getId(), splits, true)));
+            for (int partitionId : ImmutableSet.copyOf(assignedSplits.keySet())) {
+                // remove the entry from the collection to let GC reclaim the memory
+                Set<ScheduledSplit> splits = assignedSplits.removeAll(partitionId);
+                TaskSource taskSource = new TaskSource(tableScan.getId(), splits, true);
+                SerializedPrestoSparkTaskSource serializedTaskSource = new SerializedPrestoSparkTaskSource(serializeCompressed(taskSourceCodec, taskSource));
+                taskSourcesMap.put(partitionId, serializedTaskSource);
+            }
         }
 
+        long allTaskSourcesSerializedSizeInBytes = taskSourcesMap.values().stream()
+                .mapToLong(serializedTaskSource -> serializedTaskSource.getBytes().length)
+                .sum();
+        log.info("Total serialized size of all task sources for fragment %s: %s", fragmentId, DataSize.succinctBytes(allTaskSourcesSerializedSizeInBytes));
+
         List<List<SerializedPrestoSparkTaskSource>> taskSourcesByPartitionId = new ArrayList<>();
```

Comment:
We can get rid of `ListMultimap<Integer, SerializedPrestoSparkTaskSource> taskSourcesMap`  and use `List<List<SerializedPrestoSparkTaskSource>> taskSourcesByPartitionId` directly to further optimize memory.

Reply:
The container itself shouldn't use a lot of memory. There shouldn't be more than 10k - 40k partitions in a worst case. Although a hash map uses slightly more memory that shouldn't be a big issue.